python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -63,3 +67,252 @@ def get_instance_manager_state(
+                new_status=new_instance_status,
+            )
+        instance.status = new_instance_status
+        InstanceUtil._record_status_transition(instance, new_instance_status, details)
+
+    @staticmethod
+    def _record_status_transition(
+        instance: Instance, status: Instance.InstanceStatus, details: str
+    ):
+        """Records the status transition.
+
+        Args:
+            instance: The instance to update.
+            status: The new status to transition to.
+        """
+        now_ns = time.time_ns()
+        instance.status_history.append(
+            Instance.StatusHistory(
+                instance_status=status,
+                timestamp_ns=now_ns,
+                details=details,
+            )
+        )
+
+    @staticmethod
+    def get_valid_transitions() -> Dict[
+        "Instance.InstanceStatus", Set["Instance.InstanceStatus"]
+    ]:
+        return {
+            # This is the initial status of a new instance.
+            Instance.QUEUED: {
+                # Cloud provider requested to launch a node for the instance.
+                # This happens when the a launch request is made to the node provider.
+                Instance.REQUESTED
+            },
+            # When in this status, a launch request to the node provider is made.
+            Instance.REQUESTED: {
+                # Cloud provider allocated a cloud node for the instance.
+                # This happens when the cloud node first appears in the list of running
+                # cloud nodes from the cloud node provider.
+                Instance.ALLOCATED,
+                # Cloud provider fails to allocate one. Either as a timeout or
+                # the launch request fails immediately.
+                Instance.ALLOCATION_FAILED,
+            },
+            # When in this status, the cloud node is allocated and running. This
+            # happens when the cloud node is present in node provider's list of
+            # running cloud nodes.
+            Instance.ALLOCATED: {
+                # Ray needs to be install and launch on the provisioned cloud node.
+                # This happens when the cloud node is allocated, and the autoscaler
+                # is responsible for installing and launching ray on the cloud node.
+                # For node provider that manages the ray installation and launching,
+                # this state is skipped.
+                Instance.RAY_INSTALLING,
+                # Ray is already installed and running on the provisioned cloud node.
+                # This happens when a ray node joins the ray cluster, and the instance
+                # is discovered in the set of running ray nodes from the Ray cluster.
+                Instance.RAY_RUNNING,
+                # Instance is requested to be stopped, e.g. instance leaked.
+                # This happens when the autoscaler is terminating the ray process on
+                # the instance, e.g. idle termination.
```

Comment:
This should be RAY_STOPPING? I feel this comment is not right.

Reply:
yeah, good call